### PR TITLE
[GOBBLIN-1418] Add config to make it possible to specify a default Authenticator for a job

### PR DESCRIPTION
**Pull request by jack-moseley**
… a job

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1418

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

Setting the default java Authenticator allows you to run a job and force all proxy/http calls to go through this Authenticator. This is useful for example for a job where going through a proxy (which requires user/password) is desired, but using APIs with underlying http clients that don't support providing user/password. (See [stackoverflow](https://stackoverflow.com/questions/1626549/authenticated-http-proxy-with-java/43263674) discussion about this).  This PR makes this possible by:

- Adding a class extending Authenticator which provides a user/password from job properties (with decryption)
- Adding an optional key that will set the default Authenticator when the job launches (in AbstractJobLauncher)

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Successfully ran a job that was not working by providing user/password directly in jvm args

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-api/src/main/java/org/apache/gobblin/password/EncryptedPasswordAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Properties;
+
+import com.google.common.base.Preconditions;
+
+
+/**
+ * {@link Authenticator} that uses a username and password from the provided {@link Properties} to authenticate, and
+ * also decrypts the password using {@link PasswordManager}
+ */
+public class EncryptedPasswordAuthenticator extends Authenticator {
```
Comment (sv2000) [inline gobblin-api/src/main/java/org/apache/gobblin/password/EncryptedPasswordAuthenticator.java:31]: Can we add unit tests for this class?

Comment (jack-moseley) [reply gobblin-api/src/main/java/org/apache/gobblin/password/EncryptedPasswordAuthenticator.java:31]: Is there anything worth testing here? The only actual logic is calling PasswordManager, and based on PasswordManagerTest, it seems like we can't even enable tests that use that due to flakiness.

**Comment by jack-moseley** [comment]
Added this logic to the `MRJobLauncher` as well, since otherwise this would not apply to MR tasks launched by the job

File: gobblin-api/src/main/java/org/apache/gobblin/password/EncryptedPasswordAuthenticator.java (hunk 1 above)
Comment (jack-moseley) [reply gobblin-api/src/main/java/org/apache/gobblin/password/EncryptedPasswordAuthenticator.java:31]: Added a disabled test that checks an encrypted password, and an enabled test that just uses a password with no encryption.

**Review by sv2000 — APPROVED** [review]
+1. LGTM. For future PRs, please avoid doing a "git push -f", since it is hard to know the changes since the last review.

**Pull request closed without merging**